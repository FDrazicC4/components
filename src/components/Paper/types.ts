import React from 'react';
import { StyledComponent } from '@components/types';


export type Variants = 'elevation' | 'outlined';


export interface Props extends StyledComponent {

    /**
     * The content of the component.
     */
    children?: React.ReactNode,

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     * @default 'div'
     */
    component?: React.ElementType,

    /**
     * Shadow depth, corresponds to `dp` in the spec.
     * It accepts values between 0 and 24 inclusive.
     * @default 1
     */
    elevation?: number,

    /**
     * If `false`, rounded corners are disabled.
     * @default true
     */
    isRounded?: boolean,

    /**
     * The variant to use.
     * @default 'elevation'
     */
    variant?: Variants,

}
