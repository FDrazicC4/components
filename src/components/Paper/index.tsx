import React from 'react';
import clsx from 'clsx';
import { Props } from './types';


/**
 * The `Paper` component serves as a wrapping element that provides a card-like or paper-like appearance.
 */
export const Paper = React.forwardRef(({
    className,
    component: Component = 'div',
    isRounded = true,
    elevation = 1,
    variant = 'elevation',
    ...other
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const classes = clsx(
        'brew-Paper',
        `brew-Paper--elevation-${String(elevation)}`,
        `brew-Paper--variant-${variant}`,
        { 'brew-Paper--isRounded': isRounded },
        className,
    );


    return <Component className={classes} ref={ref} {...other} />;


});


export default Paper;
