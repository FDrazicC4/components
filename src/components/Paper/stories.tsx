import React from 'react';
import { select, boolean, number } from '@storybook/addon-knobs';
import Paper from './index';
import Flag from '@components/Flag';
import { Variants } from './types';


export default {
    component: Paper,
    parameters: {
        componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Stable|Surfaces/Paper',
};


const styles = { height: '100px', margin: '8px', width: '100px' };


export const General = (): React.ReactElement => {

    const elevation: number = number('elevation', 1);
    const isRounded: boolean = boolean('isRounded', true);
    const variant: Variants = select('variant', ['elevation', 'outlined'], 'elevation');

    return (
        <div style={{ display: 'flex', flexWrap: 'wrap' }}>
            <Paper elevation={elevation} isRounded={isRounded} style={styles} variant={variant} />
            <Paper elevation={elevation + 1} isRounded={isRounded} style={styles} variant={variant} />
            <Paper elevation={elevation + 2} isRounded={isRounded} style={styles} variant={variant} />
        </div>
    );

};


export const Default = (): React.ReactElement => <Paper>Default</Paper>;
export const Elevations = (): React.ReactElement => (
    <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        <Paper elevation={0} style={styles}>0</Paper>
        <Paper elevation={1} style={styles}>1</Paper>
        <Paper elevation={2} style={styles}>2</Paper>
        <Paper elevation={3} style={styles}>3</Paper>
        <Paper elevation={4} style={styles}>4</Paper>
        <Paper elevation={5} style={styles}>5</Paper>
    </div>
);
export const Outlined = (): React.ReactElement => (
    <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        <Paper elevation={0} style={styles} variant="outlined">0</Paper>
        <Paper elevation={1} style={styles} variant="outlined">1</Paper>
        <Paper elevation={2} style={styles} variant="outlined">2</Paper>
        <Paper elevation={3} style={styles} variant="outlined">3</Paper>
        <Paper elevation={4} style={styles} variant="outlined">4</Paper>
        <Paper elevation={5} style={styles} variant="outlined">5</Paper>
    </div>
);
export const Boxed = (): React.ReactElement => (
    <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        <Paper elevation={0} isRounded={false} style={styles}>0</Paper>
        <Paper elevation={1} isRounded={false} style={styles}>1</Paper>
        <Paper elevation={2} isRounded={false} style={styles}>2</Paper>
        <Paper elevation={3} isRounded={false} style={styles}>3</Paper>
        <Paper elevation={4} isRounded={false} style={styles}>4</Paper>
        <Paper elevation={5} isRounded={false} style={styles}>5</Paper>
    </div>
);

