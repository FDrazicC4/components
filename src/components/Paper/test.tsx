import React from 'react';
import renderer from 'react-test-renderer';
import Paper from './index';


describe('Paper', () => {

    it('renders correctly with defaults', () => {
        const tree = renderer
            .create(<Paper>Foo</Paper>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with properties', () => {
        const tree = renderer
            .create((
                <Paper
                    className="foo-bar"
                    component="span"
                    elevation={2}
                    isRounded={false}
                    style={{ width: '100px' }}
                    variant="outlined"
                >
                    Foo
                </Paper>
            ))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

});
