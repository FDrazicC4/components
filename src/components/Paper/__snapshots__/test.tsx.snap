// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paper renders correctly with defaults 1`] = `
<div
  className="brew-Paper brew-Paper--elevation-1 brew-Paper--variant-elevation brew-Paper--isRounded"
>
  Foo
</div>
`;

exports[`Paper renders correctly with properties 1`] = `
<span
  className="brew-Paper brew-Paper--elevation-2 brew-Paper--variant-outlined foo-bar"
  style={
    Object {
      "width": "100px",
    }
  }
>
  Foo
</span>
`;
